File: samples/MicroServices/NBB.Invoices.FSharp/NBB.Invoices.FSharp/Data/Invoice.fs
namespace NBB.Invoices.FSharp.Data

open System
open NBB.Invoices.FSharp.Domain
open InvoiceAggregate

module InvoiceRepoImpl =
    let handle<'a> : InvoiceRepository.SideEffect<'a> -> 'a =
        function
        | InvoiceRepository.GetById (invoiceId, continuation) ->
            { Id = invoiceId
              ClientId = Guid.NewGuid()
              ContractId = Guid.NewGuid() |> Some
              Amount = 100m
              PaymentId = None }
            |> continuation
        | InvoiceRepository.Save (invoice, continuation) -> continuation ()

File: samples/MicroServices/NBB.Invoices/NBB.Invoices.Domain/InvoiceAggregate/Invoice.cs
// Copyright (c) TotalSoft.
// This source code is licensed under the MIT license.

using System;
using NBB.Domain;

namespace NBB.Invoices.Domain.InvoiceAggregate
{
    public class Invoice : EventedAggregateRoot<Guid>
    {
        public Guid InvoiceId { get; private set; }

        public Guid ClientId { get; private set; }

        public Guid? ContractId { get; private set; }

        public decimal Amount { get; private set; }

        public bool IsPayed => PaymentId.HasValue;

        public Guid? PaymentId { get; private set; }

        //needed 4 repository should be private
        public Invoice()
        {

        }

        public Invoice(Guid clientId, Guid? contractId, decimal amount)
        {
            InvoiceId = Guid.NewGuid();
            Amount = amount;
            ClientId = clientId;
            ContractId = contractId;

            AddEvent(new InvoiceCreated(InvoiceId, amount, clientId, contractId));
        }

        public override Guid GetIdentityValue() => InvoiceId;

        public void MarkAsPayed(Guid paymentId, Guid? contractId)
        {
            PaymentId = paymentId;
            AddEvent(new InvoicePayed(InvoiceId, paymentId, contractId));
        }
    }
}
